addons/stock/report/product_label_report.py

```diff
@@ -22,11 +29,43 @@ def _get_report_values(self, docids, data):
         quantity_by_product = defaultdict(list)
         for p, q in data.get('quantity_by_product').items():
             product = Product.browse(int(p))
-            quantity_by_product[product].append((product.barcode, q))
+            product_info = {
+                'barcode': product.barcode,
+                'quantity': q,
+                'display_name_markup': markupsafe.Markup(product.display_name),
+                'default_code': self._get_markup_default_code(product)
+            }
+            quantity_by_product[product].append(product_info)
         if data.get('custom_barcodes'):
             # we expect custom barcodes to be: {product: [(barcode, qty_of_barcode)]}
             for product, barcodes_qtys in data.get('custom_barcodes').items():
-                quantity_by_product[Product.browse(int(product))] += (barcodes_qtys)
+                product = Product.browse(int(product))
+                new_barcodes_qtys = []
+                for barcode_qty in barcodes_qtys:
+                    new_barcodes_qtys.append({
+                        'barcode': barcode_qty[0],
+                        'quantity': barcode_qty[1],
+                        'display_name_markup': markupsafe.Markup(product.display_name),
+                        'default_code': self._get_markup_default_code(product)
+                    }
+                    )
+                quantity_by_product[product] += new_barcodes_qtys
```

Comment:
```suggestion
                for barcode_qty in barcodes_qtys:
                    quantity_by_product[product].append({
                        'barcode': barcode_qty[0],
                        'quantity': barcode_qty[1],
                        'display_name_markup': markupsafe.Markup(product.display_name),
                        'default_code': self._get_markup_default_code(product)
                    })
```
Since in the end we join `quantity_by_product[product]` and `new_barcodes_qtys` anyway, I don't think it's necessary to do that extra step.